File: P165 Compare Version Numbers/solution.py
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        version1_weight_sum = self._calculate_weight_sum(version1)
        version2_weight_sum = self._calculate_weight_sum(version2)
        if version1_weight_sum > version2_weight_sum:
            return 1
        elif version1_weight_sum < version2_weight_sum:
            return -1
        return 0
        
        
    def _calculate_weight_sum(self, version: str) -> int:
        numbers_str = version.split(".")
        weight_sum = 0
        weight = 10^250
        for number_str in numbers_str:
            weight_sum += int(number_str) * weight
            weight /= 10
            
        return weight_sum

File: P165 Compare Version Numbers/Solution.java
class Solution {
    public int compareVersion(String version1, String version2) {
        double version1Weight = calculateWeightSum(version1);
        double version2Weight = calculateWeightSum(version2);
        if(version1Weight > version2Weight) {
            return 1;
        } else if(version1Weight < version2Weight) {
            return -1;
        } else {
            return 0;
        }
    }
    
    private static double calculateWeightSum(String version) {
        String[] numbersStr = version.split("\\.");
        double weightSum = 0;
        double weight = Math.pow(10, 250);
        for(int i = 0; i < numbersStr.length; i++) {
            System.out.print(numbersStr[i] + " ");
            System.out.println(weight);
            weightSum += Integer.valueOf(numbersStr[i]) * weight;
            weight /= 10;
        }
         return weightSum;
    }
}